### ISSUE TITLE
Use the paste crate for building the name of the class struct in glib_wrapper!

### ISSUE THREAD

**Issue by sdroege** [gtk-rs/glib#707]
See https://crates.io/crates/paste

Currently we require to pass in `FooClass` (or `FooInterface`) to `glib_wrapper!` only for being able to define a struct with that name. That's rather redundant and can be solved by using the `paste` crate.

For that to work we will also not to (hidden) re-export the `paste` macro from glib but that shouldn't be a problem really.

@ids1024 Maybe something you'd be interested to look into?

**Comment by sdroege** [comment]
Also when doing that it would probably make sense to keep the old variants of the macro around for a while to not immediately break all code, and remove them in a few weeks.

It would also need a small update to [gir](https://github.com/gtk-rs/gir/).

**Comment by sdroege** [comment]
@GuillaumeGomez Wanted to write a proc-macro inside glib-macros for this specific use-case, i.e. appending a `Class` to an existing identifier.

**Comment by ids1024** [comment]
It would be nice to not have to specify the class struct name, while also not silently bringing another type into scope. Not that I can think of a good way to do both...

**Comment by ids1024** [comment]
Hm, is there any reason the class struct couldn't be a generic type?

`glib_object_wrapper!` has this case:

```rust
    (@class_impl $name:ident, $ffi_class_name:ty, $rust_class_name:ident) => {
        #[repr(C)] 
        #[derive(Debug)]
        pub struct $rust_class_name($ffi_class_name);
              
        unsafe impl $crate::object::IsClassFor for $rust_class_name {
            type Instance = $name;
        }
                
        unsafe impl Send for $rust_class_name { }
        unsafe impl Sync for $rust_class_name { }
    };
```

Could this be generic with something like `pub struct ClassWrapper<T>(T)`, instantiated as `ClassWrapper<$ffi_class_name>`? I suppose the trouble is that `IsClassFor` couldn't be implemented generically, and the compiler would reject `impl IsClassFor for ClassWrapper<Object>` from another crate... But there's probably some way to adjust the use of types and traits to something that will work.

If this could be made to work, it wouldn't be necessary to specify a new identify (it would just be an instance of a generic).

**Comment by sdroege** [comment]
There's also the problem that other crates will want to implement extensions for their class struct, like [here](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/blob/ba719ac90c1266637f0ab3254fc1f24bc9020139/gstreamer/src/subclass/element.rs#L300-339) in GStreamer. Making that work with a generic type is going to be challenging.

**Comment by ids1024** [comment]
> Making that work with a generic type is going to be challenging.

I don't see what the issue is there, as long as GStreamer is both defining the trait and implementing it. Rust crates are always free to implement their own traits on instantiations of foreign generic types.

**Comment by sdroege** [comment]
Indeed, that should be fine. So maybe that actually works. But please put the Rust type as generic parameter, not the FFI type. The FFI types should not appear anywhere unless needed.

So maybe something like `pub struct Class<T: SomeTrait>(T::Class)`. Want to give it a try? Probably needs a second variant for interfaces.

**Comment by ids1024** [comment]
At first glance, this appears to be valid:

```rust
#[repr(C)]
pub struct Class<T: ObjectType>(T::GlibClassType);

unsafe impl<T: ObjectType> IsClassFor for Class<T> {
    type Instance = T;
}

unsafe impl<T: ObjectType> Send for Class<T> { }
unsafe impl<T: ObjectType> Sync for Class<T> { }
```

A further refactoring would eliminate the `IsClassFor` trait, and just make anything related to that use the struct.

> Also when doing that it would probably make sense to keep the old variants of the macro around for a while to not immediately break all code, and remove them in a few weeks.

Regardless of whether it's done through a generic type or `paste`, this is a bit awkward since it means temporarily ending up with twice as many variants of the macro...

Side note: the class struct should probably be [`#[repr(transparent)]`](https://doc.rust-lang.org/nomicon/other-reprs.html#reprtransparent) rather than `#[repr(C)]`, right? Assuming that's supported by the oldest Rust version gtk-rs is targetting.

**Comment by sdroege** [comment]
> Regardless of whether it's done through a generic type or `paste`, this is a bit awkward since it means temporarily ending up with twice as many variants of the macro...

Indeed, so maybe let's do it all at once. That means a) PR for GLib, b) PR for gir for the autogenerated code, c) once all those are merged all the other crates have to be updated/regenerated, c) the CI passes again.

Want to give it a try?

> A further refactoring would eliminate the IsClassFor trait, and just make anything related to that use the struct.

Let's do that at the same time then so that everything only has to be changed once :)

> Side note: the class struct should probably be [`#[repr(transparent)]`](https://doc.rust-lang.org/nomicon/other-reprs.html#reprtransparent) rather than `#[repr(C)]`, right? Assuming that's supported by the oldest Rust version gtk-rs is targetting.

That's ok to use now. Back then repr(C) was used because repr(transparent) was not available yet. In practice both will behave the same in the cases where we use it for a wrapper struct, but let's better be more correct.

Another such case is `glib::Value` and `glib::SendValue`. Can't remember any others right now.

**Comment by ids1024** [comment]
> Want to give it a try?

I have some initial code for this, though getting the `Deref`/`DerefMut` implementations  to work is slightly complicated. I have a solution introducing another trait, and am thinking about a way to simplify that.

**Comment by sdroege** [comment]
Oh yeah, I can see how the `Deref` part could become tricky here.

**Comment by ids1024** [comment]
`Deref` can't be implemented in the macro, or other crates using the macro get the error about implementing a foreign trait on a foreign type.

My current solution, which seems to work, is an additional marker trait to specify what it should `Deref` to:

```rust
pub unsafe trait SuperClassIs: ObjectType {
    type Super: ObjectType;
}

impl<T: SuperClassIs> ops::Deref for Class<T> {
    type Target = Class<T::Super>;
    ...
}
```

Then that trait can just be implemented in the macro where `Deref` is currently implemented.

I tried moving `type Super` here to the `ObjectType` trait itself, which is possible with some changes to the macro. But then you get `reached the recursion limit while auto-dereferencing`, since `Class<Object>` derefs to `Class<Object>`. So I guess it isn't possible for every instance of `ObjectType` to deref to another `ObjectType`.

So I suppose the additional trait is the right solution, modulo perhaps better naming.

**Comment by sdroege** [comment]
Yeah that makes sense. Elsewhere we call it "parent (class)" and not "super" so that's better kept consistent. Could maybe do `ParentClassIs` or `IsSubclassOf`. I don't have any better suggestions :)
